### PR TITLE
remove a giant problem in settingshelper, it was loading the same set…

### PR DESCRIPTION
**Pull request by benjaoming**
…tings many times, assertion doesnt work

#3810

Conflicts:
	python-packages/fle_utils/settingshelper.py

### CODE REVIEW

**Comment by benjaoming** [comment]
I'm pretty sure this works because my local 0.13 branch is still working, but I don't understand where Travis icons are on this PR?

**Comment by benjaoming** [comment]
Now I see Travis, nevermind

**Comment by benjaoming** [comment]
After change these settings are loaded:

```
django_extensions
kalite.distributed
kalite.django_cherrypy_wsgiserver
kalite.coachreports
kalite.testing
securesync
fle_utils.chronograph
kalite.facility
kalite.main
kalite.i18n
kalite.topic_tools
kalite.contentload
kalite.i18n
kalite.main
kalite.testing.loadtesting
kalite.topic_tools
kalite.topic_tools
kalite.facility
kalite.main
kalite.main
kalite.playlist
kalite.caching
kalite.updates
kalite.control_panel
securesync
kalite.updates
kalite.facility
kalite.student_testing
kalite.topic_tools
kalite.contentload
kalite.remoteadmin
kalite.i18n
kalite.testing
kalite.ab_testing
fle_utils.chronograph
kalite.control_panel
kalite.testing
```

Before:

```
django_extensions
kalite.distributed
kalite.django_cherrypy_wsgiserver
kalite.coachreports
kalite.testing
securesync
fle_utils.chronograph
kalite.facility
kalite.main
kalite.i18n
kalite.topic_tools
kalite.contentload
kalite.i18n
kalite.main
kalite.i18n
kalite.topic_tools
kalite.contentload
kalite.testing.loadtesting
kalite.topic_tools
kalite.i18n
kalite.contentload
kalite.topic_tools
kalite.i18n
kalite.facility
securesync
fle_utils.chronograph
kalite.main
kalite.contentload
kalite.facility
securesync
fle_utils.chronograph
kalite.main
kalite.main
securesync
fle_utils.chronograph
kalite.facility
securesync
fle_utils.chronograph
kalite.main
kalite.i18n
kalite.i18n
kalite.main
securesync
fle_utils.chronograph
kalite.i18n
kalite.main
kalite.testing
securesync
fle_utils.chronograph
kalite.facility
kalite.i18n
kalite.testing.loadtesting
kalite.topic_tools
kalite.i18n
kalite.contentload
securesync
fle_utils.chronograph
kalite.topic_tools
kalite.i18n
kalite.facility
kalite.contentload
kalite.facility
kalite.facility
kalite.i18n
kalite.i18n
kalite.playlist
kalite.contentload
kalite.topic_tools
kalite.testing
securesync
fle_utils.chronograph
kalite.facility
kalite.i18n
kalite.testing.loadtesting
kalite.i18n
kalite.facility
securesync
fle_utils.chronograph
kalite.facility
kalite.testing
securesync
fle_utils.chronograph
kalite.testing.loadtesting
securesync
fle_utils.chronograph
kalite.i18n
kalite.caching
kalite.testing
securesync
fle_utils.chronograph
kalite.facility
kalite.i18n
kalite.testing.loadtesting
kalite.topic_tools
kalite.i18n
kalite.contentload
kalite.i18n
kalite.facility
securesync
fle_utils.chronograph
kalite.updates
securesync
fle_utils.chronograph
fle_utils.chronograph
kalite.topic_tools
kalite.contentload
kalite.facility
kalite.control_panel
kalite.facility
kalite.contentload
kalite.topic_tools
kalite.contentload
kalite.facility
securesync
fle_utils.chronograph
securesync
fle_utils.chronograph
kalite.updates
kalite.testing
kalite.facility
kalite.i18n
kalite.testing.loadtesting
kalite.topic_tools
kalite.i18n
kalite.contentload
fle_utils.chronograph
kalite.topic_tools
kalite.i18n
kalite.facility
kalite.contentload
kalite.facility
kalite.control_panel
kalite.facility
kalite.i18n
kalite.contentload
kalite.i18n
kalite.facility
kalite.facility
kalite.testing
kalite.topic_tools
kalite.i18n
kalite.contentload
kalite.testing.loadtesting
kalite.i18n
kalite.student_testing
kalite.topic_tools
kalite.testing
kalite.testing.loadtesting
kalite.i18n
kalite.contentload
kalite.contentload
kalite.remoteadmin
kalite.i18n
kalite.testing
kalite.testing.loadtesting
kalite.ab_testing
fle_utils.chronograph
kalite.control_panel
kalite.testing
securesync
fle_utils.chronograph
kalite.facility
kalite.main
kalite.i18n
kalite.topic_tools
kalite.contentload
kalite.i18n
kalite.main
kalite.i18n
kalite.topic_tools
kalite.contentload
kalite.testing.loadtesting
kalite.topic_tools
kalite.i18n
kalite.contentload
```

**Comment by benjaoming** [comment]
I'm not sure why we still have the same modules loaded several times, I would advice to just merge this one and nevermind about the rest!

I'm deprecating the dreaded settingshelper for 0.14 anyways.

**Comment by MCGallaspy** [comment]
I would rather not merge anything into 0.13 at this point, given that we plan to release a new version soon anyway.

**Comment by benjaoming** [comment]
Well if it's good just merge it in, I don't think there will be a point release on 0.13 right now anyways, and even if there will never be one, it's a harmless merge.

I just don't like obvious backports to be lying around forgotten.

**Comment by MCGallaspy** [comment]
Are we planning on backporting anything? My understanding is that we're not going to release any more `0.13.x` installers, so I would rather see bugfixes go into `develop` for 0.14.x. I think merging something into an older release branch and not a newer one (either by merging `0.13.x` into `develop` again or opening the same PR targeting `develop`) creates the potential to "fix" bugs in an older version that remain extant in newer versions.

**Comment by MCGallaspy** [comment]
What I mean is, I'd like `develop` and future release branches to strictly be fast-forwards of `0.13.x`. But if that's a silly expectation, please let me know.

**Comment by benjaoming** [comment]
I dunno what the strategy should be in general, I just when with this strategy: "this is a one-line cherry-pick that fixes an important error" :)

**Comment by MCGallaspy** [comment]
Sadly, it seems to cause some test failures :(
How about we get it working on `0.13.x` and then merge up into `develop` from there?

File: python-packages/fle_utils/settingshelper.py [hunk 1]
```diff
@@ -68,9 +68,7 @@ def import_installed_app_settings(installed_apps, global_vars, cur_app="__root__
                 #logging.warn("(%s) %s is already set; resetting can cause confusion." % (app, var))
                 pass
 
-        #print "\n%s" % processed_apps
-        processed_apps = processed_apps.union(set([app]))
-        #print processed_apps
+        processed_apps.add(app)
```
Comment (aronasorman) [inline python-packages/fle_utils/settingshelper.py]: Can you explain why this change fixed this problem?

**Comment by aronasorman** [comment]
Yeah I'd like to close 0.13 at this point. 0.14 is close.

File: python-packages/fle_utils/settingshelper.py (hunk 1 above)
Comment (benjaoming) [inline python-packages/fle_utils/settingshelper.py]: It's explained over in #3811

Comment (benjaoming) [inline python-packages/fle_utils/settingshelper.py]: "I think the case here is that the way set.union works is by reference, and set.add works by value... well, that's my theory, I didn't look it up because it empirically just works for me."

**Comment by benjaoming** [comment]
I don't understand why a one-liner for a previous branch that doesn't have an imminent release is getting closed? Why not just leave it open for later, in case a new 0.13.x is released?

**Comment by aronasorman** [comment]
@benjaoming ok, I see what you mean that we *might* have another 0.13 release. In fact if we're reopening this then let's just merge this then, but not treat this single change as critical enough to warrant another patch release for 0.13.

**Comment by benjaoming** [comment]
Nono, that was never the intention :)

**Comment by MCGallaspy** [comment]
So what's the plan then for #3811? One-liner though it may be, it causes test failures in both `0.13.x.` and `develop`. Shall we merge #3811 and deal with the consequences?
